### DISCUSSION TITLE
DuckDB and Fugue integration

### DISCUSSION THREAD

**Discussion by goodwanghan** [duckdb/duckdb#2447] [Ideas]
Hello,

I am the creator of [Fugue](https://github.com/fugue-project/fugue). It builds a SQL like language on top of standard SQL with extra syntax, so people can use SQL to describe their data pipelines end to end and it can invoke python extensions when SQL can't do the job.

I found DuckDB is a great SQL solution on top of the dataframe concepts, it fits perfectly with the Fugue SQL framework. So I have added DuckDB as a SQL backend of Fugue with just [a few lines of code](https://github.com/fugue-project/fugue/blob/master/fugue_duckdb/execution_engine.py) also with the help from @Mytherin . I have to say, it is so fast on dataframes, much faster than Sqlite, and even faster than Fugue's built in SQL solution (which directly translates SQL to pandas operations).

The only problem I see is related with this [issue](https://github.com/duckdb/duckdb/issues/2446), and with the 0.3.0 offlical version, we can't get correct result for a complex data pipeline with multiple selects. But the master branch has fixed the issues, that works well.

I have created a [Kaggle notebook](https://www.kaggle.com/goodwanghan/fugue-duckdb) to demonstrate the integration of Fugue and DuckDB. In this case if not using `duck` but using `sqlite` and `qpd` as SQL backend, it will take very long time to finish. I have some ideas to make the integration even faster (avoid conversion to arrow or pandas in intermediate steps). But in general it is already amazing.

Thanks
Han

**Comment by hannes** [comment]
Could you maybe tell us about your ideas to make this even faster?

**Comment by goodwanghan** [comment]
So I may be able to wrap the relation as a Fugue dataframe, most of its functions plus the execution engine functions can be done with relation. For some certain cases such as using extensions, it can be converted to arrow or pandas dataframes.

In this way we expect in most of the time the intermediate dataframe is either virtual or is only in duckdb system. So we minimized unnecessary dataframe conversions.
